### PR TITLE
Sitl debugging, some bug fixes and various cleanup

### PR DESCRIPTION
**Pull request by CarloWood**
This used to be https://github.com/PX4/Firmware/pull/5432

I renamed the branch to sitl_debugging because I think the per-target optimization control is main subject in this all.

Some documentation written for this can be reviewed here:
https://github.com/CarloWood/Devguide/blob/d8c8bd418e6fcef71432a5511a5416e264f5c1f9/simulation-debugging.md

Please see the commit messages for a detailed description of what this does.

### CODE REVIEW

**Comment by CarloWood** [comment]
Is there anything else that I should do before this can be merged?

**Comment by bkueng** [comment]
The `list_cmake_targets` target does not work properly for me, as it outputs:
```
[1/1] List of cmake targets that ...e matched by PX4_NO_OPTIMIZATION:
drivers__boards__sitl\ndrivers__device\ndrivers__gps\ndrivers__pwm_out_sim\ndrivers__sf0x__sf0x_tests\nexamples__attitude_estimator_ekf\nexamples__ekf_att_pos_estimator\nexamples__mc_att_control_multiplatform\nexamples__mc_pos_control_multiplatform\nexamples__px4_simple_app\nlib__DriverFramework__framework\nlib__controllib\nlib__conversion\nlib__ecl\nlib__external_lgpl\nlib__geo\nlib__geo_lookup\nlib__launchdetection\nlib__mathlib\nlib__mathlib__math__filter\nlib__rc\nlib__rc__rc_tests\nlib__runway_takeoff\nlib__tailsitter_recovery\nlib__terrain_estimation\nmodules__attitude_estimator_q\nmodules__commander\nmodules__commander__commander_tests\nmodules__controllib_test\nmodules__dataman\nmodules__ekf2\nmodules__fw_att_control\nmodules__fw_pos_control_l1\nmodules__land_detector\nmodules__local_position_estimator...
```
If I replace `echo` with `printf` it works. 
Also, the list gets appended every time cmake re-runs, so I get something like this:
```
[1/1] List of cmake targets that ...e matched by PX4_NO_OPTIMIZATION:
drivers__boards__sitl
drivers__boards__sitl
drivers__boards__sitl
drivers__boards__sitl
drivers__device
drivers__device
drivers__device
drivers__device
drivers__gps
drivers__gps
drivers__gps
drivers__gps
drivers__pwm_out_sim
drivers__pwm_out_sim
drivers__pwm_out_sim
drivers__pwm_out_sim
...
```

**Comment by CarloWood** [comment]
Ah, good catch thanks! I'll fix that later today.

**Comment by CarloWood** [comment]
Is 'printf' POSIX? Does 'echo -e' work for you? It's kinda weird that it works for me and not for you, I wonder where is the difference. Looks like mine gobbles up a backslash more.

**Comment by bkueng** [comment]
Yes `printf` is posix (http://pubs.opengroup.org/onlinepubs/9699919799/utilities/printf.html), but `echo -e` works too. Not sure where the difference comes from, maybe it's `ninja`?

**Comment by CarloWood** [comment]
The merge conflict is trivial - but how can I do that when I'm not supposed to simply merge master into my branch and push it again?

**Comment by LorenzMeier** [comment]
A rebase on master would do the trick.

**Comment by bkueng** [comment]
You missed to replace echo with printf for the `list_cmake_targets` target. Other than that it looks good.
If you rebase & force push the CI retest.

File: Makefile [hunk 1]
```diff
@@ -158,7 +158,7 @@ nuttx_% posix_% qurt_%:
 # Abbreviated config targets.
 
 # nuttx_ is left off by default.
-tap-% asc-% px4fmu-% px4-% mindpx-% crazyflie_default:
+tap-% asc-% px4fmu-% px4-% mindpx-% crazyflie_%:
```
Comment (LorenzMeier) [inline Makefile]: The underscore here looks sort of wrong - are we sure crazyflie is actually using one?

Comment (LorenzMeier) [inline Makefile]: Never mind, that's correct.

**Comment by bkueng** [comment]
rebased & applied

**Comment by kd0aij** [comment]
@CarloWood I tried "make posix_sitl_ekf2 jmavsim___gdb" and got no connection to QGC.
There is also no response to ctrl-C: (Ubuntu 15.10)
```
12:21:00 {(d2626d7...)} ~/git/kd0aij/PX4/Firmware$ make posix_sitl_ekf2 jmavsim___gdb
PX4 CONFIG: /home/markw/git/kd0aij/PX4/Firmware/build_posix_sitl_ekf2
ninja: Entering directory `/home/markw/git/kd0aij/PX4/Firmware/build_posix_sitl_ekf2'
[1/1] cd /home/markw/git/kd0aij/PX4/Firmware/build_posix_s... /home/markw/git/kd0aij/PX4/Firmware/build_posix_sitl_ekf2
args: /home/markw/git/kd0aij/PX4/Firmware/build_posix_sitl_ekf2/src/firmware/posix/px4 posix-configs/SITL/init/ekf2 gdb jmavsim none /home/markw/git/kd0aij/PX4/Firmware /home/markw/git/kd0aij/PX4/Firmware/build_posix_sitl_ekf2
SITL ARGS
sitl_bin: /home/markw/git/kd0aij/PX4/Firmware/build_posix_sitl_ekf2/src/firmware/posix/px4
rcS_dir: posix-configs/SITL/init/ekf2
debugger: gdb
program: jmavsim
model: none
src_path: /home/markw/git/kd0aij/PX4/Firmware
build_path: /home/markw/git/kd0aij/PX4/Firmware/build_posix_sitl_ekf2
empty model, setting iris as default
SITL COMMAND: /home/markw/git/kd0aij/PX4/Firmware/build_posix_sitl_ekf2/src/firmware/posix/px4 /home/markw/git/kd0aij/PX4/Firmware /home/markw/git/kd0aij/PX4/Firmware/posix-configs/SITL/init/ekf2/iris
GNU gdb (Ubuntu 7.10-1ubuntu2) 7.10
Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/markw/git/kd0aij/PX4/Firmware/build_posix_sitl_ekf2/src/firmware/posix/px4...Buildfile: /home/markw/git/kd0aij/PX4/Firmware/Tools/jMAVSim/build.xml
done.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
data path: /home/markw/git/kd0aij/PX4/Firmware
commands file: /home/markw/git/kd0aij/PX4/Firmware/posix-configs/SITL/init/ekf2/iris
[New Thread 0x7ffff6d47700 (LWP 18156)]
18446744073709551586 WARNING: setRealtimeSched failed (not run as root?)
[New Thread 0x7ffff7ff4700 (LWP 18157)]
[New Thread 0x7ffff6546700 (LWP 18158)]
[New Thread 0x7ffff7fea700 (LWP 18159)]
[New Thread 0x7ffff5d45700 (LWP 18160)]
[Thread 0x7ffff7fea700 (LWP 18159) exited]
[New Thread 0x7ffff7fe0700 (LWP 18161)]
[New Thread 0x7ffff5544700 (LWP 18162)]
[Thread 0x7ffff7fe0700 (LWP 18161) exited]

______  __   __    ___ 
| ___ \ \ \ / /   /   |
| |_/ /  \ V /   / /| |
|  __/   /   \  / /_| |
| |     / /^\ \ \___  |
\_|     \/   \/     |_/

px4 starting.

[Thread 0x7ffff7ff4700 (LWP 18157) exited]
[New Thread 0x7ffff7fd6700 (LWP 18163)]
[New Thread 0x7ffff4d43700 (LWP 18164)]
[Thread 0x7ffff7fd6700 (LWP 18163) exited]
INFO  [dataman] Unkown restart, data manager file 'rootfs/fs/microsd/dataman' size is 103090 bytes

make_dirs:

compile:

create_run_jar:
[New Thread 0x7ffff7fcc700 (LWP 18165)]
[Thread 0x7ffff7fcc700 (LWP 18165) exited]
[New Thread 0x7fffeffff700 (LWP 18166)]
INFO  [platforms__posix__drivers__ledsim] LED::init
INFO  [platforms__posix__drivers__ledsim] LED::init
INFO  [simulator] Not using /dev/ttyACM0 for radio control input. Assuming joystick input via MAVLink.
INFO  [simulator] Waiting for initial data on UDP port 14560. Please start the flight simulator to proceed..

copy_res:

BUILD SUCCESSFUL
Total time: 0 seconds
Options parsed, starting Sim.
Starting GUI...
3D [dev] 1.6.0-pre12-daily-experimental daily

Init MAVLink
INFO  [simulator] Got initial simuation data, running sim..
[New Thread 0x7fffef7fe700 (LWP 18217)]
[New Thread 0x7ffff7fb5700 (LWP 18218)]
[New Thread 0x7fffeeffd700 (LWP 18219)]
[Thread 0x7ffff7fb5700 (LWP 18218) exited]
INFO  [pwm_out_sim] MODE_8PWM
[New Thread 0x7ffff7fab700 (LWP 18220)]
[New Thread 0x7fffee7fc700 (LWP 18221)]
[Thread 0x7ffff7fab700 (LWP 18220) exited]
[New Thread 0x7ffff7fa1700 (LWP 18222)]
[Thread 0x7ffff7fa1700 (LWP 18222) exited]
[New Thread 0x7fffedffb700 (LWP 18223)]
[New Thread 0x7ffff7f97700 (LWP 18224)]
[New Thread 0x7fffed7fa700 (LWP 18225)]
[Thread 0x7ffff7f97700 (LWP 18224) exited]
INFO  [tone_alarm] startup
[New Thread 0x7fffecff9700 (LWP 18226)]
[New Thread 0x7ffff7f8d700 (LWP 18227)]
[Thread 0x7ffff7f8d700 (LWP 18227) exited]
[New Thread 0x7fffcbfff700 (LWP 18228)]
[New Thread 0x7ffff7f83700 (LWP 18229)]
[New Thread 0x7fffcb7fe700 (LWP 18230)]
[Thread 0x7ffff7f83700 (LWP 18229) exited]
[New Thread 0x7ffff7f79700 (LWP 18231)]
[New Thread 0x7fffcaffd700 (LWP 18232)]
[Thread 0x7ffff7f79700 (LWP 18231) exited]
[New Thread 0x7ffff7f6f700 (LWP 18233)]
[New Thread 0x7fffca7fc700 (LWP 18234)]
[Thread 0x7ffff7f6f700 (LWP 18233) exited]
[New Thread 0x7ffff7f65700 (LWP 18235)]
[New Thread 0x7fffc9ffb700 (LWP 18236)]
[Thread 0x7ffff7f65700 (LWP 18235) exited]
INFO  [mavlink] mode: Normal, data rate: 4000000 B/s on udp port 14556 remote port 14550
[New Thread 0x7fffc97fa700 (LWP 18237)]
[New Thread 0x7ffff7f5b700 (LWP 18238)]
[New Thread 0x7fffc8ff9700 (LWP 18239)]
[Thread 0x7ffff7f5b700 (LWP 18238) exited]
INFO  [mavlink] mode: Onboard, data rate: 4000000 B/s on udp port 14557 remote port 14540
[New Thread 0x7fffabfff700 (LWP 18240)]
[New Thread 0x7ffff7f51700 (LWP 18241)]
[New Thread 0x7fffab7fe700 (LWP 18242)]
[Thread 0x7ffff7f51700 (LWP 18241) exited]
INFO  [sdlog2] [blackbox] rootfs/fs/microsd/log/2016-09-13
WARN  [sdlog2] sdlog2: failed setting sched params
[New Thread 0x7fffaaffd700 (LWP 18243)]
INFO  [sdlog2] [blackbox] recording: 18_21_12.px4log
^C^C^C^C^C
```

**Comment by CarloWood** [comment]
@kd0aij make posix_sitl_ekf2 jmavsim___gdb works for me. I don't see anything being compiled when you issue the command though. Is your problem perhaps fixed by doing a 'make clean' first? And/or deleting the build directory... 

Anyway, I don't see how the problem that you describe can be related to this patch, as it only changes the build system a bit (ie, the Makefile). Possible expected problems would be of the kind where targets are not built, or give some kind of error when building.

**Comment by kd0aij** [comment]
"make clean" before "make posix_sitl_ekf2 jmavsim___gdb" still results in no communication with QGC.
I also get no output from "make list_vmd_make_targets". Perhaps that provides a clue as to what's wrong.

**Comment by CarloWood** [comment]
Can you please remove /home/markw/git/kd0aij/PX4/Firmware/build_posix_sitl_ekf2
And then copy&paste the full output of your make command, including the make command and the pwd that you run it from, on a paste site, like I did?

**Comment by kd0aij** [comment]
07:00:10 (master) ~/git/kd0aij/PX4/Firmware$ ninja --version
1.6.0

https://gist.github.com/kd0aij/8b0fae37a42f520dbd459b3e706b8b53

**Comment by CarloWood** [comment]
06:53:31 {master} ~/git/kd0aij/PX4/Firmware$ make list_cmake_targets 
06:53:36 {master} ~/git/kd0aij/PX4/Firmware$ make list_vmd_make_targets

Those are build directory targets, not targets of the root Makefile.
By the way, after my patch of #5491 you'll get:

    sean:/usr/src/px4/Firmware/Firmware.git>make list_cmake_targets
    Makefile:292: *** "list_cmake_targets cannot be the first argument. Use 'make help|list_config_targets' to get a list of all possible [configuration] targets.".  Stop.

You can run this in the build directory, or you could run:
    $ make posix_sitl_ekf2 list_cmake_targets
    $ make posix_sitl_ekf2 list_vmd_make_targets

I don't know why you're having problems, can't really think of a reason.
However, I see you have PX4_NO_OPTIMIZATION set (why?).
Do things work when you unset that environment variable?

**Comment by kd0aij** [comment]
Thanks, Carlo.
Regarding PX4_NO_OPTIMIZATION, I was following the instructions here:
https://github.com/CarloWood/Devguide/blob/d8c8bd418e6fcef71432a5511a5416e264f5c1f9/simulation-debugging.md#compiler-optimization

Unsetting it seemed to make no difference; QGC doesn't load parameters when
the simulator starts. Do I need to issue any gdb commands first?
https://gist.github.com/kd0aij/cccbecd574ff481d278080898e0d6deb

Also, how would I start the gdb session from within Eclipse (or attach it)?
That would be much preferable to the command line interface.

**Comment by CarloWood** [comment]
> Regarding PX4_NO_OPTIMIZATION, I was following the instructions here:

That's not an instruction is it? It's an example, saying IF you'd do this then the effect is that.

I cannot help you with Eclipse, I never used that. I'm fairly new to this whole project myself, so I'm afraid I can't help. What you pasted looks like it works to my eyes, and certainly the part that is related to my commit. I'd appreciate it if you could take the time to do a git bisect to find out which commit is causing this; are you 100% sure it is caused by my changes?

**Comment by kd0aij** [comment]
Do you use QGC with SITL?

**Comment by bkueng** [comment]
@kd0aij I just tested, I can connect QGC to jmavsim with gdb (`make posix_sitl_ekf2 jmavsim___gdb`)

**Comment by CarloWood** [comment]
After discussing this in gitter with Mark, it became apparent that this is probably unrelated to the changes made in this PR: it's simply the first time that Mark tries to use gdb. Nevertheless, with his help I was able to reproduce the issue and I'm willing to investigate it; after all, having a working connection between QGC and the firmware while debugging in gdb helps ;).

So far the following came out of the investigation:
Things work when gdb is not used, and something locks up when gdb is used. This even happens when gdb is attached afterwards.
Not only px4 becomes unresponsive (read: appears to be locked up), even gdb itself is often locked up (ie, doesn't react to control-C).
Sending the px4 process SIGINT signals causes it to continue shortly; ie QGC reacts (sees the process) and I get a pxh> prompt, but immediately after things return to the locked-in state.

**Pull request closed without merging**